{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Interacting with KEGG data using Biopython"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##Â Introduction\n",
    "\n",
    "Biopython 1.65 was [released](http://news.open-bio.org/news/2014/12/biopython-1-65-released/) on 17th December 2014. One of the new additions to the libraries was the `Bio.Graphics.KGML` module, which can be used for customised renderings of KEGG metabolic pathways. \n",
    "\n",
    "This notebook is an introduction to the `Bio.KEGG` and `Bio.Graphics.KGML` modules, providing examples of usage, including data retrieval, manipulation and output rendering."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running cells in this notebook\n",
    "\n",
    "This is an interactive notebook, which means you are able to run the code that is written in each of the cells.\n",
    "\n",
    "<div class=\"alert alert-info\" role=\"alert\">\n",
    "To run the code in a cell, you should:\n",
    "<ol>\n",
    "<li>Place your mouse cursor in the cell, and click (this gives the cell <i>focus</i>) to make it active\n",
    "<li>Hold down the <b>Shift</b> key, and press the <b>Return<b> key.\n",
    "</ol>\n",
    "</div>\n",
    "\n",
    "If this is successful, you should see the input marker to the left of the cell change from\n",
    "\n",
    "```\n",
    "In [ ]:\n",
    "```\n",
    "\n",
    "to (for example)\n",
    "\n",
    "```\n",
    "In [1]:\n",
    "```\n",
    "\n",
    "and you may see output appear below the cell."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0. Setting up"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We need to import Biopython and parts of the `KEGG` and `KGML` modules, but first we'll check that a suitably recent version of Biopython is installed (1.65 or higher for Python 2; 1.67 or higher for Python 3):\n",
    "\n",
    "<div class=\"alert alert-danger\" role=\"alert\">\n",
    "For <b>Biopython<=1.67</b>, you will need to run this notebook with a <b>Python2 kernel</b>!\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import Bio\n",
    "Bio.__version__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We're going to be using some other Biopython modules, and working with images in the notebook, so we also prepare for this, with the following imports."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.KEGG.REST import *\n",
    "from Bio.KEGG.KGML import KGML_parser\n",
    "from Bio.Graphics.KGML_vis import KGMLCanvas\n",
    "from Bio.Graphics.ColorSpiral import ColorSpiral\n",
    "\n",
    "from IPython.display import Image, HTML\n",
    "\n",
    "import os\n",
    "import random"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To visualise the PDF output generated by this library inline, and to provide a `head`-like peek at text output, the functions below are defined:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# A bit of code that will help us display the PDF output\n",
    "def PDF(filename):\n",
    "    return HTML('<iframe src=%s width=700 height=350></iframe>' % filename)\n",
    "\n",
    "# A bit of helper code to shorten long text\n",
    "def head(text, lines=10):\n",
    "    \"\"\"Print the first lines lines of the passed text.\"\"\"\n",
    "    print('\\n'.join(text.split('\\n')[:lines] + ['[...]']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Output files will be stored in the subdirectory `kgml_output`, which may need to be created:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Create subdirectory to store output\n",
    "outdir = \"kgml_output\"\n",
    "try:\n",
    "    os.makedirs(outdir, exist_ok=True)\n",
    "except TypeError:\n",
    "    if not os.path.isdir(outdir):\n",
    "        os.mkdir(outdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Retrieving data from KEGG with `Bio.KEGG.REST`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The [KEGG](http://www.genome.jp/kegg/) database provides a RESTful interface to its data (see [http://www.kegg.jp/kegg/rest/](http://www.kegg.jp/kegg/rest/)), which we can use via Biopython's `Bio.KEGG.REST` module.\n",
    "\n",
    "The KEGG interface is not as well documented as some other resources (such as NCBI or Ensembl), and KEGG does not provide any usage guidelines. To avoid risking overloading the service, Biopython restricts us to three calls per second. Be warned also that the conditions of service include:\n",
    "\n",
    ">\"This service should not be used for bulk data downloads\".\n",
    "\n",
    "**NOTE:** The KEGG API returns data in `Byte` handles (in Python3, these are strings prefixed with the `b` character). Biopython's interface converts text data to `String` handles, but leaves image data as the `Byte` type."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1a. Obtaining information about a database with `kegg_info()`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can obtain basic information about the KEGG databases (e.g. \"pathway\", \"kegg\", \"hsa\", etc.) using the `kegg_info()` function. This returns a handle which can be read as plaintext, with some useful information about the resource."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Information about the Kyoto Encyclopedia of Genes and Genomes\n",
    "print(kegg_info(\"kegg\").read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Information about the KEGG Pathway Database\n",
    "print(kegg_info(\"pathway\").read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Information about the Escherichia coli K-12 MG1655 KEGG Genes Database\n",
    "print(kegg_info(\"eco\").read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1b. Obtaining a list of entries in a database with `kegg_list()`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `kegg_list()` function retrieves a handle, which provides a plaintext tab-separated `key:value` list of entry identifiers and definitions in the given database or set of database entries. \n",
    "\n",
    "The `kegg_list()` function allows the organism to be specified for a database, with an optional second argument."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# List all pathways in the pathway database\n",
    "head(kegg_list('pathway').read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Only list pathways present in E. coli K-12 MG1655\n",
    "head(kegg_list('pathway', 'eco').read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If only the organism is specified, then a list of gene entries is returned."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# E. coli K-12 MG1655 genes\n",
    "head(kegg_list('eco').read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The KEGG API also allows the `+` operator to be used to combine list identifiers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Compound entry C01290 and glycan entry G00092\n",
    "print(kegg_list('C01290+G00092').read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "But `kegg_list()` and other functions will take a Python `list` of identifiers, and combine them automatically:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Compound entry C01290 and glycan entry G00092\n",
    "print(kegg_list(['C01290', 'G00092']).read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1c. Finding a specific entry with `kegg_find()`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `kegg_find(database, query)` function finds entries with matching `query` keywords or other `query` data in a given `database`. Data is returned as tab-separated `key:value` plain text."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Find shiga toxin genes\n",
    "head(kegg_find('genes', 'shiga+toxin').read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Find shiga toxin genes only in Escherichia coli O111 H-11128 (EHEC)\n",
    "print(kegg_find('eoi', 'shiga+toxin').read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The API lets us query on useful properties, such as compound molecular weight:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Compounds with molecular weight between 300 and 310g/mol\n",
    "head(kegg_find('compound', '300-310/mol_weight').read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1d. Retrieving specific entries with `kegg_get()`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `kegg_get()` function lets us retrieve data from KEGG in a range of different formats. \n",
    "\n",
    "Database entries can be retrieved directly as plain text; sequences may be returned as database entries, amino acid or nucleotide sequences in FASTA format; compounds as database entries, or images (`.gif`); pathways as database entries, KGML, or images (`.png`)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compounds can be retrieved as their database entry in plain text, or as a `.gif` image."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Compound as database entry\n",
    "head(kegg_get(\"cpd:C01290\").read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Compound as image\n",
    "Image(kegg_get(\"cpd:C01290\", \"image\").read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Gene data can be retrieved as the plain text database entry, or a FASTA nucleotide or protein sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Gene as database entry\n",
    "head(kegg_get(\"ece:Z5100\").read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Gene as amino acid sequence\n",
    "print(kegg_get(\"ece:Z5100\", \"aaseq\").read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Gene as nucleotide sequence\n",
    "print(kegg_get(\"ece:Z5100\", \"ntseq\").read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Parsing a returned sequence with SeqIO, and converting format\n",
    "seq = SeqIO.read(kegg_get(\"ece:Z5100\", \"ntseq\"), 'fasta')\n",
    "print(seq.format('stockholm'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pathways can be returned as database entries in plain text, or in the KGML format, or as `.png` images. These last two formats are especially useful for generating custom pathway renderings with `Bio.Graphics.KGML`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Pathway as database entry\n",
    "head(kegg_get(\"hsa05130\").read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Pathway as image (png)\n",
    "Image(kegg_get(\"hsa05130\", \"image\").read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Pathway as KGML\n",
    "head(kegg_get(\"hsa05130\", \"kgml\").read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Rendering pathways with `Bio.KEGG` and `Bio.Graphics.KGML`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2a. Retrieving renderings from KEGG"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The easiest way to render a pathway is to retrieve an image file directly from KEGG with `kegg_get()`, as above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Render central metabolism\n",
    "Image(kegg_get(\"map01100\", \"image\").read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Render fatty-acid biosynthesis\n",
    "Image(kegg_get(\"map00061\", \"image\").read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "KEGG also provides renderings that are specific to an organism, indicating which elements of the pathway are present in their database for that organism."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For central metabolism, compounds and reactions are only coloured if they are present in the specified organism. Otherwise, they are rendered in grey."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Render E.coli K-12 MG1655 central metabolism\n",
    "Image(kegg_get(\"eco01100\", \"image\").read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With other pathways, the reactions and/or genes are coloured green when they are present in the organism, and in white otherwise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Render E.coli K-12 MG1655 fatty-acid biosynthesis\n",
    "Image(kegg_get(\"eco00061\", \"image\").read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Three types of reference pathway can also be rendered: `ko`, `ec`, and `rn`. These are shown with blue boxes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Render reference fatty-acid biosynthesis\n",
    "#Image(kegg_get(\"ko00061\", \"image\").read())\n",
    "#Image(kegg_get(\"ec00061\", \"image\").read())\n",
    "Image(kegg_get(\"rn00061\", \"image\").read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2b. Reproducing KEGG renderings from KGML"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "KEGG provides descriptions of their pathway maps in an exchange format called KGML ([spec](http://www.kegg.jp/kegg/xml/)). This doesn't quite match the maps that can be downloaded through the public interface. If you're a subscriber to KEGG, then you have access to an alternative set of [KGML+](http://www.kegg.jp/kegg/download/Readme/README.kgml+) files. The Biopython interface uses the public API, so we only work with KGML files."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To obtain a pathway's KGML file from KEGG, we can use the `kegg_get()` function. In the cell below, we store the downloaded data for two maps in the variables `ko_map` and `eco_map`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Get KGML for fatty-acid biosynthesis\n",
    "ko_map = (kegg_get(\"ko00061\", \"kgml\").read())    # KO version (KEGG orthologues)\n",
    "eco_map = (kegg_get(\"eco00061\", \"kgml\").read())  # E. coli version"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Visualising the contents of the KGML file, it can be seen that two useful kinds of elements are described in the file: `ortholog` and `compound` entries. Reactions and other connections are usually not defined directly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# View the contents of ko_map KGML\n",
    "head(ko_map, 20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Typically, `ortholog`s and `compound`s are the only items that can be manipulated on a KEGG pathway map. KGML does not describe the complete map, with its connectors and labels, as provided in the downloaded image."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The base maps (such as, in this case, `map00061`) do not contain `ortholog` or `compound` features, and consequently there is no associated KGML file. Attempting to retrieve KGML data for these maps produces an `HTTP 404: not found` error."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Trying to retrieve base map KGML generates an HTTP 404: not found error\n",
    "#base_map = (kegg_get(\"map00061\", \"kgml\").read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The beautiful images generated by KEGG are rendered as `.png` files, which may be downloaded, and the features described in the accompanying KGML description overlaid, with custom formatting. This is what is currently done in the majority of cases, here."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To do this, first we need to parse the KGML we get from KEGG into a `Pathway` object. This holds an internal representation of the pathway information, such as `ortholog`s and `compound`s (and, as we'll see later, `reaction`s and relations.\n",
    "\n",
    "The `Pathway` object provides some useful functions for manipulation of elements in the KGML representations, and allows us to write out a new KGML file that preserves the changes we make. Using `print` on the `Pathway` object gives us some summary information."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`KGML_parser` will read a KGML file on disk, and the stream returned from a `kegg_get()` function, as below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pathway = KGML_parser.read(kegg_get(\"ko00061\", \"kgml\"))\n",
    "print(pathway)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To render this data as an image, we need to set up a `KGMLCanvas` object, from `Bio.Graphics.KGML_vis`, to draw on. We do this by instantiating a `KGMLCanvas` object with our pathway data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "canvas = KGMLCanvas(pathway)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have an image, we can render it. At the moment, there is only the capability to render to `.pdf`. To render to a file, call the `draw()` method of the canvas, with the output filename as an argument.\n",
    "\n",
    "Using the cell below, we also view the `.pdf` file that is created in an `iFrame`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "canvas.draw(os.path.join(outdir, \"fab_map.pdf\"))\n",
    "PDF(os.path.join(outdir, \"fab_map.pdf\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "But there's something wrong with this picture. It shows only the elements defined in the KGML file. All the beautiful KEGG layout, including the lines indicating the flow of the biochemistry, is missing. That's because the KGML file doesn't contain any information about it, directly. \n",
    "\n",
    "However, each `Pathway` object created from KGML file has an `image` attribute with the URL of its `.png` format representation. To use this as a background, all we need to do to render this is specify `import_imagemap=True` when creating the canvas, or set `canvas.import_imagemap = True` once the canvas is created. (It is also possible to specify an image file on disk as the background.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "canvas.import_imagemap = True\n",
    "canvas.draw(os.path.join(outdir, \"fab_map_with_image.pdf\"))\n",
    "PDF(os.path.join(outdir, \"fab_map_with_image.pdf\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This could be encapsulated in a useful helper function, as in the cell below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def draw_kegg_map(map_id, outdir):\n",
    "    \"\"\"Render a local PDF of a KEGG map with the passed map ID.\"\"\"\n",
    "    # Get the background image first\n",
    "    pathway = KGML_parser.read(kegg_get(map_id, \"kgml\"))\n",
    "    canvas = KGMLCanvas(pathway, import_imagemap=True)\n",
    "    img_filename = \"%s.pdf\" % map_id\n",
    "    canvas.draw(os.path.join(outdir, img_filename))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2c. Customised renderings from KGML - Part 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the `Bio.Graphics` model of rendering KGML, we are superimposing the elements described in a KGML file over an associated background image. As a result, only the elements that are described in the KGML file (or that we introduce ourselves...) can be modified. \n",
    "\n",
    "The current interface to identifying and retrieving pathway elements is a little obscure as of Biopython 1.65 - if you want to retrieve a particular element by ID or annotation -  but it is possible to modify the graphical elements that are rendered, to produce customised maps from KGML data.\n",
    "\n",
    "The modifications we can make are:\n",
    "\n",
    "* shape (`graphics.type`)\n",
    "* location ($x,y$ co-ordinates: `graphics.x`, `graphics.y`)\n",
    "* height (`graphics.height`)\n",
    "* width (`graphics.width`)\n",
    "* foreground colour (`graphics.fgcolor`)\n",
    "* background colour (`graphics.bgcolor`)\n",
    "* label (`graphics.name`)\n",
    "\n",
    "but this still gives us a lot of flexibility in terms of visual annotation from a dataset. We can also choose to display, or not, elements of interest, and transparency effects are available."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### A motivating example"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As a motivating example, we're going to use the fatty-acid biosynthesis pathway (`map00061`) from earlier, and make the following modifications:\n",
    "\n",
    "1. colour the `ortholog` elements according to a random choice of colours\n",
    "2. change the size of each `compound` element to reflect its molecular mass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We saw the default rendering of the `ko00061` pathway earlier:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pathway = KGML_parser.read(kegg_get(\"ko00061\", \"kgml\"))\n",
    "canvas = KGMLCanvas(pathway, import_imagemap=True)\n",
    "canvas.draw(os.path.join(outdir, \"fab_map_with_image.pdf\"))\n",
    "PDF(os.path.join(outdir, \"fab_map_with_image.pdf\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2c.1 Changing `ortholog` background colours"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can get access to all the pathway `ortholog` elements *via* the `pathway.orthologs` attribute. This will list all `ortholog` elements in the pathway as `Entry` objects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pathway.orthologs[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "print(pathway.orthologs[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each `Entry` object has a `graphics` attribute that stores information about how the item is rendered, in a list of `Graphics` objects - this reflects the data structure in the KGML representation, which may have more than one graphical entity per entry in the pathway."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pathway.orthologs[0].graphics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each `Graphics` object has attributes that may be modified directly:\n",
    "\n",
    "    name         Label for the graphics object\n",
    "    x            X-axis position of the object (int)\n",
    "    y            Y-axis position of the object (int)\n",
    "    coords       polyline co-ordinates, list of (int, int) tuples\n",
    "    type         object shape\n",
    "    width        object width (int)\n",
    "    height       object height (int)\n",
    "    fgcolor      object foreground color (hex RGB)\n",
    "    bgcolor      object background color (hex RGB)\n",
    "    \n",
    "There are also two read-only properties that are provided for convenience\n",
    "\n",
    "    bounds       returns the graphical bounds of the object\n",
    "    centre       returns the centre of the object as an (x, y) tuple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "element = pathway.orthologs[0].graphics[0]\n",
    "attrs = [\"name\", \"x\", \"y\", \"coords\", \"type\", \"width\", \"height\",\n",
    "         \"fgcolor\", \"bgcolor\", \"bounds\", \"centre\"]\n",
    "for a in attrs:\n",
    "    print(\"element.{0: <10}:\\t{1}\".format(a, getattr(element, a)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To change the `ortholog` colours, it's possible to loop over the `ortholog` elements in the pathway, and modify their `graphics.bgcolor` attributes.\n",
    "\n",
    "As KGML expects hex colours in its specification, we also provide hex colours here, using the helper function below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Helper function to convert colour as RGB tuple to hex string\n",
    "def rgb_to_hex(rgb):\n",
    "    rgb = tuple([int(255*val) for val in rgb])\n",
    "    return '#' + ''.join([hex(val)[2:] for val in rgb]).upper()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this example we're defining arbitrary colours, but some obvious uses might include the use of colour to represent some biologically-significant value, like transcript abundance, $\\frac{dN}{dS}$, or flux through that step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Define arbitrary colours\n",
    "colorspiral = ColorSpiral()\n",
    "colorlist = colorspiral.get_colors(len(pathway.orthologs))\n",
    "\n",
    "# Change the colours of ortholog elements\n",
    "for color, element in zip(colorlist, pathway.orthologs):\n",
    "    for graphic in element.graphics:\n",
    "        graphic.bgcolor = rgb_to_hex(color)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "canvas = KGMLCanvas(pathway, import_imagemap=True)\n",
    "canvas.draw(os.path.join(outdir, \"fab_map_new_colours.pdf\"))\n",
    "PDF(os.path.join(outdir, \"fab_map_new_colours.pdf\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2c.2 Changing `compound` sizes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can get access to the pathway `compound` elements *via* `pathway.compounds`, just as for `ortholog`s, and to change sizes, we can loop over them, modifying both `width` and `height`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As with the colour changes, the sizes used are arbitrary, but this kind of modification (maybe incorporating colour changes, too) could be used to represent biochemical measurements such as metabolite concentration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Change the sizes of compound elements\n",
    "for size, element in zip(range(8, 8+len(pathway.compounds)), pathway.compounds):\n",
    "    for graphic in element.graphics:\n",
    "        graphic.width = size\n",
    "        graphic.height = size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "canvas = KGMLCanvas(pathway, import_imagemap=True)\n",
    "canvas.draw(os.path.join(outdir, \"fab_map_new_sizes.pdf\"))\n",
    "PDF(os.path.join(outdir, \"fab_map_new_sizes.pdf\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2d. Customised renderings from KGML - Part 2: central metabolism"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The maps we've been considering so far have not been completely described by the publicly-available KGML file. This is true for the great majority of KEGG maps. But there are three KEGG maps: `ko01100`, `ko01110`, and `ko01120` for which this is not the case, and the KGML file provides a lot more information. Specifically, they also describe lines representing reactions that connect metabolite compounds. This gives us quite a lot of scope for modifying their presentation to convey useful information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# The three KEGG maps with lines representing reactions.\n",
    "maps = ['ko01100', 'ko01110', 'ko01120']\n",
    "[draw_kegg_map(map, outdir) for map in maps]\n",
    "print(kegg_get(maps).read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "PDF(os.path.join(outdir, maps[0]+'.pdf'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "PDF(os.path.join(outdir, maps[1]+'.pdf'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "PDF(os.path.join(outdir, maps[2]+'.pdf'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Unlike the custom renderings that can be obtained from KEGG directly, with this module we now have full control over line width, element size, element colour, and labelling options.\n",
    "\n",
    "To avoid noise in the image, we can set `import_imagemap=False` when making the canvas or drawing to a file for these maps."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, let's set each reaction line to a random value between 1 and 10. This could be adapted to represent carbon flux measurements, for example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Use the bacterial diverse environments map\n",
    "pathway = KGML_parser.read(kegg_get(\"ko01120\", \"kgml\"))\n",
    "\n",
    "# Change the widths of reaction entries elements\n",
    "for element in pathway.orthologs:\n",
    "    for graphic in element.graphics:\n",
    "        graphic.width = random.randrange(1, 10, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "canvas = KGMLCanvas(pathway, import_imagemap=False)\n",
    "canvas.draw(os.path.join(outdir, \"bacteria_mod_widths.pdf\"))\n",
    "PDF(os.path.join(outdir, \"bacteria_mod_widths.pdf\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can modify pathway colours in the same way as before, except that now we have to modify the foreground, not the background colour, through the attribute `fgcolor`. The foreground colour should be specified as a hex string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Define arbitrary colours\n",
    "colorspiral = ColorSpiral()\n",
    "colorlist = colorspiral.get_colors(len(pathway.orthologs))\n",
    "\n",
    "# Change the colours of ortholog elements\n",
    "for color, element in zip(colorlist, pathway.orthologs):\n",
    "    for graphic in element.graphics:\n",
    "        graphic.fgcolor = rgb_to_hex(color)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "canvas = KGMLCanvas(pathway, import_imagemap=False)\n",
    "canvas.draw(os.path.join(outdir, \"bacteria_mod_colour.pdf\"))\n",
    "PDF(os.path.join(outdir, \"bacteria_mod_colour.pdf\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  },
  "widgets": {
   "state": {},
   "version": "1.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
